[Messenger] Batch delivery

hgraca opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix [#47828](https://github.com/symfony/symfony/issues/47828)
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/18566


Currently, the doctrine transport only delivers one message at a time,
resulting in low performance and a very high amount of hits to the DB.
(one hit per message)

With batch delivery, the transport will retrieve several messages in a single query.

fabpot commented on src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/Connection.php:
```diff
@@ -46,10 +48,13 @@ class Connection implements ResetInterface
         'auto_setup' => true,
     ];
 
+    public const FLAGGED_FOR_DELETION = '9999-12-31 23:59:59';
```
To be reverted as not related to this PR.

fabpot replied:
Or at least, make it private.

hgraca replied:
I made it private. 
What about [this one](https://github.com/symfony/symfony/pull/51034/files#r1281056793)?
Is it ok to have that one public, or should I drop that commit? (only ahs that change)

hgraca replied:
I removed this and other small fixes from the PR as to remove noise in the PR

fabpot commented on src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/Connection.php:
```diff
@@ -66,6 +71,8 @@ class Connection implements ResetInterface
     protected $queueEmptiedAt;
     private ?SchemaSynchronizer $schemaSynchronizer;
     private bool $autoSetup;
+    /** @var array<int> */
```
```suggestion
    /** @var int[] */
```

hgraca replied:
changed

hgraca commented on src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/DoctrineTransportFactory.php:
```diff
@@ -20,9 +21,12 @@
  */
 class DoctrineTransportFactory implements TransportFactoryInterface
 {
+    public const DSN_PREFIX = 'doctrine://';
```
After reading [this comment](https://github.com/symfony/symfony/pull/51034#pullrequestreview-1554567967), I'm doubting if this public constant ok, or if I should drop this commit? (only has this change)

fabpot replied:
I would drop the constant altogether.

hgraca replied:
Done, and tkx for the CR btw :)

hgraca commented:
I've rebased this PR on branch 6.4, and removed all the commits that were merely small improvements (I will open PRs for them later).

I see that the pipeline is failing but it seems to be unrelated to this PR.

I am, however, afraid that this PR might be getting stale since it is open for a few weeks now.

Is this normal or is there something else that I should do? (I'm new to contributing to Symfony so I really don't know)

hgraca commented:
Rebased again and put the changelog in the correct file

hgraca committed:
Add batch delivery to Messenger Doctrine transport

Currently, the doctrine transport only delivers one message at a time,
resulting in low performance and a very high amount of hits to the DB.
(one hit per message)

With batch delivery, the transport will retrieve several messages
in a single query.

Co-authored-by: Alexander Malyk <shu.rick.ifmo@gmail.com>

ro0NL commented on src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/DoctrineTransport.php:
```diff
@@ -37,6 +39,18 @@ public function __construct(Connection $connection, SerializerInterface $seriali
+    public function __destruct()
+    {
+        // The worker using this transport might have pulled 50 msgs out of the mq, marking them as "in process",
+        // but because of its options (ie --limit, --failure-limit, --memory-limit, --time-limit) it might terminate
```
shouldnt you read the options instead :thinking:

hgraca replied:
Then I think we would increase complexity and it wouldn't scale.

For `--limit` we could do some calculations... would be doable despite the increase in complexity..
But for `-failure-limit` we would need to somehow keep track of how many failures happened, probably replicating logic that is already built somewhere. For `--memory-limit` and `--time-limit` the same.

Then, if/when in the future we would add more options, someone would need to also adjust this code, to be able to handle the new option.

With the implementation in this PR however, we don't need to do anything special to adjust to a particular option, neither now nor in the future if/when we add new options.

ro0NL replied:
got ya, i was thinking we might correlate limit and batch_size somehow, but i agree it increases complexity for little gain, and the conceptual issue remains either way

ro0NL replied:
im not sure if `__destruct` is the best extension point compared to eg. events, but if it works .. :shrug:

stof commented on src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/Connection.php:
```diff
@@ -510,4 +541,53 @@ private function fallBackToForUpdate(QueryBuilder $query, string $sql): string
+     */
+    public function extractDoctrineEnvelopeListIds(array $doctrineEnvelopeList): array
+    {
+        return array_map(
```
`return array_column($doctrineEnvelopeList, 'id')` will do the same and might be a bit faster by keeping all the logic in the C layer.

stof commented on src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/Connection.php:
```diff
@@ -151,7 +155,7 @@ public function send(string $body, array $headers, int $delay = 0): string
     }
 
-    public function get(): ?array
+    public function get(): array
```
I suggest document the type of the returned array.

stof commented on src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/Connection.php:
```diff
@@ -200,40 +204,45 @@ public function get(): ?array
+                ],
+                [
+                    Types::DATETIME_IMMUTABLE,
+                    class_exists(ArrayParameterType::class) ? ArrayParameterType::INTEGER : Types::SIMPLE_ARRAY,
```
This is wrong. `Types::SIMPLE_ARRAY` is something totally different. The fallback for older DBAL versions is `Connection::PARAM_INT_ARRAY`.

And the code below does not check for older versions, so it is inconsistent.

stof commented on src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/DoctrineReceiver.php:
```diff
@@ -58,11 +60,16 @@ public function get(): iterable
         }
 
-        if (null === $doctrineEnvelope) {
+        if (null === $doctrineEnvelopeList) {
```
This won't ever be true. `Connection::get` cannot return `null` anymore.

stof commented on src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/DoctrineReceiver.php:
```diff
@@ -83,6 +90,11 @@ public function reject(Envelope $envelope): void
         }
     }
 
+    public function undeliverNotHandled(): void
```
should it be marked as `@internal` ?

OskarStark commented:
Closing due to lack of feedback. Feel free to reopen. Thanks for proposing.
